clarity/contracts/arkadiko-stacker-payer-v3-3.clar

```diff
@@ -0,0 +1,143 @@
+      )
+      (ok true)
+    )
+  )
+)
+
+(define-public (set-stx-redeemable (ustx-amount uint))
+  (let (
+    (dao-address (contract-call? .arkadiko-dao get-dao-owner))
+  )
+    (asserts! (is-eq contract-caller dao-address) (err ERR-NOT-AUTHORIZED))
+
+    (ok (var-set stx-redeemable ustx-amount))
+  )
+)
+
+(define-public (add-stx-redeemable (auction-id uint))
+  (let (
+    (auction (contract-call? .arkadiko-auction-engine-v4-2 get-auction-by-id auction-id))
+    (vault (contract-call? .arkadiko-vault-data-v1-1 get-vault-by-id (get vault-id auction)))
+    (difference (if (> (get total-collateral-sold auction) (get stacked-tokens vault))
+      (- (get total-collateral-sold auction) (get stacked-tokens vault))
+      u0
+    ))
+  )
+    (asserts! (not (has-vault-redeemed (get vault-id auction))) (err ERR-VAULT-ALREADY-REDEEMED))
+    (asserts! (get auction-ended vault) (err ERR-AUCTION-RUNNING))
+    (asserts! (get is-liquidated vault) (err ERR-NOT-LIQUIDATED))
+    (asserts! (is-eq (get collateral-token vault) "xSTX") (err ERR-WRONG-COLLATERAL))
+
+    (map-set vaults-redeemed { vault-id: (get vault-id auction) } { redeemed: true })
+    (ok (var-set stx-redeemable (+ (var-get stx-redeemable) difference)))
+  )
+)
+
+(define-public (redeem-stx (ustx-amount uint))
+  (let (
+    (sender tx-sender)
+    (amount (min-of ustx-amount (var-get stx-redeemable)))
+  )
+    (asserts! (is-enabled) (err ERR-EMERGENCY-SHUTDOWN-ACTIVATED))
+    (asserts! (> amount u0) (ok true))
+
+    (try! (contract-call? .arkadiko-dao burn-token .xstx-token amount sender))
+    (try! (contract-call? .arkadiko-stx-reserve-v1-1 request-stx-to-auto-payoff amount))
+    (try! (as-contract (stx-transfer? amount tx-sender sender)))
+
+    (ok (var-set stx-redeemable (- (var-get stx-redeemable) amount)))
+  )
+)
+
+(define-read-only (get-stx-redeemable-helper)
+  (let (
+    (freddie-redeemable (unwrap-panic (contract-call? .arkadiko-freddie-v1-1 get-stx-redeemable)))
+  )
+    (+ freddie-redeemable (var-get stx-redeemable))
+  )
+)
+
+(define-public (redeem-stx-helper (ustx-amount uint))
+  (let (
+    (freddie-total-redeemable (unwrap-panic (contract-call? .arkadiko-freddie-v1-1 get-stx-redeemable)))
+  )
+    (try! (contract-call? .arkadiko-freddie-v1-1 redeem-stx ustx-amount))
+
+    (if (> ustx-amount freddie-total-redeemable)
+      (redeem-stx (- ustx-amount freddie-total-redeemable))
+      (ok true)
+    )
+  )
+)
+
+(define-public (release-stacked-stx (auction-id uint))
+  (let (
+    (auction (contract-call? .arkadiko-auction-engine-v4-2 get-auction-by-id auction-id))
+  )
+    (try! (contract-call? .arkadiko-freddie-v1-1 release-stacked-stx (get vault-id auction)))
+    (try! (add-stx-redeemable auction-id))
+
+    (ok true)
+  )
+)
+
+(define-private (min-of (i1 uint) (i2 uint))
+  (if (< i1 i2)
+      i1
+      i2))
+
+
+;; Initialization
+(begin
+  ;; TODO: update for mainnet
+  (map-set vaults-redeemed { vault-id: u409 } { redeemed: true })
```

Comment:
Q: do we still need all these vaults? That list becomes pretty long and hard to fetch, but I can try 😅

Reply:
A: now that i look at it again i don't think so. Because in the method `add-stx-redeemable` we get auctions from the newest auction-engine, so it's not possible to add redeemable stx from auctions opened by the previous auction-engine.
I guess we just need to set `stx-redeemable` properly when deploying.